docs(nx-dev): add recipe for yarn pnp (#18460)
nrwl/nx: docs/shared/recipes/yarn-pnp.md

Read more about `PnP` on [Yarn's official docs](https://yarnpkg.com/features/pnp).

## Switching to Yarn v2+ (aka Berry)

When using `create-nx-workspace` with optional `--pm=yarn`, we will use the default yarn version set in your global `.yarnrc.yml` file (usually in the root of your user folder). To check what is your set yarn version, run:

```shell
 yarn --version
```

If version is in `1.x.x` range, the workspace will be created with yarn classic. To migrate to modern version of Yarn, also known as `Berry`, after the workspace has been created, run:

```shell {% path="~/workspace" %}
 yarn set version stable
```

This command will update `.yarnrc.yml` with path to local yarn bin of the correct version and will set the `packageManager` property in the root `package.json`:

```jsonc {% fileName="package.json" %}
{
  // ...
  "packageManager": "yarn@3.6.1"
}
```

## Switching to PnP

Once we are on modern version of Yarn, we can use the following command to switch to `PnP`:

```shell {% path="~/workspace" %}
 yarn config set nodeLinker pnp
```

> Your `.yarnrc.yml` will now have `nodeLinked` set to a proper method:

Reviewer: This typo needs to be fixed

```yml {% fileName=".yarnrc.yml" %}
nodeLinker: pnp
```

Once the config is changed we need to run install again:

```shell {% path="~/workspace" %}
 yarn install
```

Running install generates `.pnp.cjs` file that contains mapping of external packages and strips all the packages from the `node_modules`. The contents of your `node_modules` should now look like this:

```text
node_modules/
└── .cache
    └──  nx
```

## Dealing with inaccessible dependencies

When using Yarn Berry, Nx will set node linker to backwards compatible `node-modules` by default. This is to ensure your repository works out-of-the-box.

> Some packages come with broken dependency requirements so using strict `PnP` can lead to broken builds. This results in error similar to one below:

Reviewer: ```suggestion
Some packages come with broken dependency requirements so using strict `PnP` can lead to broken builds. This results in errors similar to the one below:
```

```shell {% path="~/workspace" command="yarn nx build my-project" %}
Error: [BABEL]: @babel/plugin-transform-react-jsx tried to access @babel/core (a peer dependency) but it isn't provided by your application; this makes the require call ambiguous and unsound.

Required package: @babel/core
```

> The problem above is something we need to fix by ensuring the requested peer dependency exists as is available:

Reviewer: ```suggestion
The problem above is something we need to fix by ensuring the requested peer dependency exists and is available:
```

```shell {% path="~/workspace" %}
 yarn add -D @babel/core
```

> Sometimes, simply adding a missing package doesn't work. In those cases the only thing we can do is to contact the author of package causing the issues and notify them of a missing dependency requirement.

Reviewer: ```suggestion
Sometimes, simply adding a missing package doesn't work. In those cases the only thing we can do is to contact the author of the package causing the issues and notify them of the missing dependency requirement.
```

> The alternative is to switch back to using `node-modules` node linker.

Reviewer: ```suggestion
The alternative is to switch back to using the `node-modules` node linker.
```
